File: python/P222.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def finddepth(self, root):
        if not root:
            return 0
        else:
            depth = 0
            root = root.right
            while root:
                depth += 1
                root = root.left
            return depth
    
    
    def countNodes(self, root):
        if not root:
            return 0
        depth = 1
        temp = root
        while temp.left:
            depth += 1
            temp = temp.left
            
        rest = 0
        cur_root = root
        cur_dep = 1
        while cur_dep < depth:
            dd = self.finddepth(cur_root)
            if dd+cur_dep == depth:
                cur_root = cur_root.right
                rest += 2**(dd-1)
                cur_dep += 1
            else:
                cur_root = cur_root.left
                cur_dep += 1
        return 2**(depth-1) + rest

File: java/P222.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int countNodes(TreeNode root) {
        if (root == null) return 0;
        int left = getheight(root.left);
        int right = getheight(root.right);
        if (left == right) {
            return countNodes(root.right) + (1<<left); // includes the whole left side and the root node total counts
        }
        return countNodes(root.left) + (1<<right);
    }
    
    
    public int getheight(TreeNode n) {
        int height = 0;
        while (n != null) {
            height++;
            n = n.left;
        }
        return height;
    }
}